equinox/nn/_embedding.py

```diff
@@ -89,3 +93,253 @@ def __call__(
                 "`eqx.nn.Embedding()(x)` should be called with a scalar index `x`. "
                 "Use `jax.vmap` if you would like to index with multiple values."
             )
+
+
+class RotaryPositionalEmbedding(Module):
+    embedding_size: int = field(static=True)
+    max_seq_len: Optional[int] = field(static=True)
+    freqs_cis: Optional[Complex[Array, "embedding_size/2"]] = field(static=True)
+
+    def __init__(
+        self,
+        embedding_size: int,
+        max_seq_len: Optional[int] = None,
+        *,
+        key: Optional[PRNGKeyArray] = None,
+        **kwargs,
+    ):
+        """**Arguments:**
+
+        `RotaryPositionalEmbedding` requires:
+
+        - `embedding_size`: Size of the token embeddings. Must be non-negative.
+        - `max_seq_len`: The maximum sequence length. Must be non-negative if provided.
+        - `key`: Not used; provided for compatibility with the rest of the Equinox API.
+            (Keyword only argument.)
+        """
+        super().__init__(**kwargs)
+        if embedding_size < 0:
+            raise ValueError("embedding_size must not be negative.")
+        if max_seq_len is not None and max_seq_len < 0:
+            raise ValueError("max_seq_len must not be negative.")
+        self.embedding_size = embedding_size
+        self.max_seq_len = max_seq_len
+
+        if self.max_seq_len is not None:
+            if (self.embedding_size, self.max_seq_len) in internal_rope_embedding_cache:
+                self.freqs_cis = internal_rope_embedding_cache[
+                    (self.embedding_size, self.max_seq_len)
+                ]
+            else:
+                self.freqs_cis = self.precompute_freqs_cis(
+                    embedding_size, self.max_seq_len
+                )
+                internal_rope_embedding_cache[
+                    (self.embedding_size, self.max_seq_len)
+                ] = self.freqs_cis
```

Comment:
You know, I think we can probably skip having this as an attribute altogether. Just create the global variable here if required, and then look it up during the forward pass.

Reply:
I'm having some problems understanding and implementing this.

I realise that it makes sense to **not** have `freqs_cis` as an attribute anymore. Instead, it's better (and uses less code - always a win) to simply check the cache during the forward pass. 

Only problem now is that we get leaked tracers!

Essentially, its like this now

```python

cache = {}

class Foo(eqx.Module):
    n: int = field(static=True)

    def __init__(self, n):
        self.n = n

    def __call__(self, x: Float[Array, "n"]):
        n, _ = x.shape

        y = None
        if n in cache:
            y = cache[n]
        else:
            y = jnp.arange(n)
            cache[n] = y #<<<<<<<<< Leaking the tracer here!
        
        return x + y
```

When JITting `Foo`, we get a leak. So, I'm not sure how to get it to work here 😞

Reply:
On second thought: do we even _really_ need to cache the `freqs_cis`? If the user calls the function with constantly changing `seq_len`s then won't Jax reJIT the function anyway?

I.e. 

```
import equinox as eqx
import jax.numpy as jnp


class Test(eqx.Module):
    def __init__(self):
        pass

    def __call__(self, x):
        print(f"JIT: {x.shape=}")
        return x + 1


t = Test()
t = eqx.filter_jit(t)


print(t(jnp.array([1])))
print(t(jnp.array([1])))

print(t(jnp.array([1, 2])))
print(t(jnp.array([1, 2, 3])))
print(t(jnp.array([1, 2])))

```

Which outputs

```
JIT: x.shape=(1,)
[2]
[2]
JIT: x.shape=(2,)
[2 3]
JIT: x.shape=(3,)
[2 3 4]
[2 3]
```

Which would suggest that Jax automatically does the caching for us, no?

Reply:
On the tracer leak: ah, try wrapping the setter in `with jax.ensure_compile_time_eval()`.

As for why we should cache this: I *believe* (but have not carefully checked) that we need to do this to work around a quirk of the XLA compiler. Namely, suppose we were to compute the (co)sines as we needed them, inside `__call__`. Now suppose we have many `RotaryPositionalEmbedding` layers, and iterated through them in a `jax.lax.scan` using the [scan-over-layers trick](https://docs.kidger.site/equinox/tricks/#improve-compilation-speed-with-scan-over-layers). Then I belive XLA is not smart enough to identify that the repeated computation can be hoisted out of the scan (i.e. to perform it only once and reuse the result every time). Instead, it will recompute them every time. Thus, it falls to us to do this manually instead.

In particular this is to do with having multiple `RotaryPositionalEmbedding` layers in the same model; it's unrelated to handling different calls with different sequence lengths.

Reply:
@Artur-Galstyan If you want something to crib off of, we have an implementation here https://github.com/stanford-crfm/levanter/blob/main/src/levanter/models/llama.py#L223 I think the only thing missing is the ensure_compile_time_eval bit.